nalgebra-sparse/src/io/matrix_market.rs

```diff
@@ -0,0 +1,861 @@
+    fn from_str(word: &str) -> Result<Self, Self::Err> {
+        match word {
+            "skew-symmetric" => Ok(StorageScheme::Skew),
+            "general" => Ok(StorageScheme::General),
+            "symmetric" => Ok(StorageScheme::Symmetric),
+            "hermitian" => Ok(StorageScheme::Hermitian),
+            _ => Err(MMError::from_kind_and_message(
+                MMErrorKind::ParsingError,
+                format!("keyword {} is unknown", word),
+            )),
+        }
+    }
+}
+
+/// Precheck if it's a valid header.
+///
+/// For more details, please check Boisvert, Ronald F., Roldan Pozo, and Karin A. Remington. The matrix market formats: Initial design. Technical report, Applied and Computational Mathematics Division, NIST, 1996.  Section 3.
+fn typecode_precheck(tc: &Typecode) -> Result<(), MMError> {
+    match tc {
+        Typecode {
+            dt: DataType::Real,
+            ss: StorageScheme::Hermitian,
+            ..
+        } => Err(MMError::from_kind_and_message(
+            MMErrorKind::InvalidHeader,
+            String::from("Real matrix can't be hermitian."),
+        )),
+        Typecode {
+            dt: DataType::Integer,
+            ss: StorageScheme::Hermitian,
+            ..
+        } => Err(MMError::from_kind_and_message(
+            MMErrorKind::InvalidHeader,
+            String::from("Integer matrix can't be hermitian."),
+        )),
+        Typecode {
+            dt: DataType::Pattern,
+            ss: StorageScheme::Hermitian,
+            ..
+        } => Err(MMError::from_kind_and_message(
+            MMErrorKind::InvalidHeader,
+            String::from("Pattern matrix can't be hermitian."),
+        )),
+        Typecode {
+            dt: DataType::Pattern,
+            ss: StorageScheme::Skew,
+            ..
+        } => Err(MMError::from_kind_and_message(
+            MMErrorKind::InvalidHeader,
+            String::from("Pattern matrix can't be skew-symmetric."),
+        )),
+        // precheck success
+        _ => Ok(()),
+    }
+}
+
+/// Base trait for matrix market types.
+pub trait MMType: Sized + Clone {
+    /// When matrix is an Integer matrix, it will convert a [i128] number to this type.
+    fn from_i128(i: i128) -> Result<Self, MMError>;
+    /// When matrix is a Real matrix, it will convert a [f64] number to this type.
+    fn from_f64(f: f64) -> Result<Self, MMError>;
+    /// When matrix is a Complx matrix, it will convert a [Complex<f64>] number to this type.
+    fn from_c64(c: Complex<f64>) -> Result<Self, MMError>;
+    /// When matrix is a Pattern matrix, it will convert a unit type [unit] to this type.
+    fn from_pattern(p: ()) -> Result<Self, MMError>;
+    /// When matrix is a Skew-symmetric matrix, it will convert itself to its negative.
+    fn negative(self) -> Result<Self, MMError>;
+    /// When matrix is a Hermitian matrix, it will convert itself to its conjugate.
+    fn conjugate(self) -> Result<Self, MMError>;
```

Comment:
`from_pattern` is a little strange... I find myself wishing that the last three methods here were just internal implementation details, rather than exposed in the trait itself. But I think this is OK for now.